README.md
**Note: The JavaScript Software Synthesizer is in active devleopment and its currently being refactored/redesigned, thus some of the features might not work as intended. For the latest stable version please visit [this repository](https://github.com/michaelkolesidis/javascript-software-synthesizer-classic).**

![JSS-01 Logo](./public/assets/logo/logo_256_white.png#gh-dark-mode-only)
![JSS-01 Logo](./public/assets/logo/logo_256.png#gh-light-mode-only)

<h1>JSS-01<br>JavaScript Software Synthesizer</h1>

[![made-with-typescript](https://img.shields.io/badge/Made%20with-TypeScript-01004c.svg)](https://developer.mozilla.org/en-US/docs/Web/JavaScript)
[![MIT license](https://img.shields.io/badge/License-AGPLv3-01004c.svg)](https://www.gnu.org/licenses/agpl-3.0.html)
[![Version](https://img.shields.io/badge/version-3.0.0-01004c.svg)](https://shields.io/)

<a href='https://ko-fi.com/michaelkolesidis' target='_blank'><img src='https://cdn.ko-fi.com/cdn/kofi6.png' style='border:0px;height:45px;' alt='Buy Me a Coffee at ko-fi.com' /></a>

## Technologies Used

<a href="https://tonejs.github.io/"><img src="https://github.com/michaelkolesidis/tech-icons/blob/main/icons/tonejs/tonejs-original.png" height="50px" /></a>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<a href="https://webmidijs.org/"><img src="https://webmidijs.org/img/webmidijs-logo-dark.svg" height="50px" /></a>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<a href="https://www.typescriptlang.org/"><img src="https://github.com/michaelkolesidis/tech-icons/blob/main/icons/typescript/typescript-original.svg" height="50px" /></a>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<a href="https://sass-lang.com/"><img src="https://github.com/michaelkolesidis/tech-icons/blob/main/icons/sass/sass-original.svg" height="50px" /></a>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<a href="https://www.yarnpkg.com/"><img src="https://github.com/michaelkolesidis/tech-icons/blob/main/icons/yarn/yarn-original.svg" height="50px" /></a>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
<a href="https://vitejs.dev/"><img src="https://vitejs.dev/logo.svg" height="50px" /></a>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;

More info about the technologies can be found <a href="https://github.com/michaelkolesidis/javascript-software-synthesizer/blob/main/README.md#technologies-used-extended">here</a>.

## Description

The _JSS-01 | JavaScript Software Synthesizer_ is a web application enabling you to make and play music in the browser. It is a [software synthesizer](https://en.wikipedia.org/wiki/Software_synthesizer) utilizing the [Web Audio API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API) by implementing a variety of JavaScript frameworks and libraries, such as [Tone.js](https://tonejs.github.io/) and [NexusUI2](https://github.com/michaelkolesidis/nexusui2). Its design is fully modular, thus it can easily be adapted and integrated into any kind of project.

My aim when creating the JSS-01 was to offer a powerful synthesizer that is simple, easy and, above all, fun to use!

<img width="900px" src="./screenshots/v.0.5.4-large.png">

More screenshots can be found <a href="https://github.com/michaelkolesidis/javascript-software-synthesizer/blob/main/README.md#screenshots">here</a>.

## Build & Run

First clone the project from GitHub:

```
git clone git@github.com:michaelkolesidis/javascript-software-synthesizer.git
cd javascript-software-synthesizer
```

Install the project dependencies:

```
yarn
```

Start the Vite development server:

```
yarn dev
```

### For versions 1.x.x

Before the introduction of Vite in version 2.0.0, after installing the dependencies, users could run _JavaScript Software Synthesizer_ locally by using a VSCode extension like **Live Server** ([Live Server on the Visual Studio Marketplace](https://marketplace.visualstudio.com/items?itemName=ritwickdey.LiveServer)) or Python's **http.server** module.

For more information, please refer to MDN's excellent article: [MDN article](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/Tools_and_setup/set_up_a_local_testing_server)

### Demo

There is an online demo available at [https://javascript-software-synthesizer.vercel.app/](https://javascript-software-synthesizer.vercel.app/)

## Introduction

Keys, knobs, sliders and buttons to tweak your sound and generate immersive soundscapes. You can also connect your MIDI keyboard and play with it!

### Quick Start

The _JSS-01 | JavaScript Software Synthesizer_ uses <strong>FM Synthesis</strong> (frequency modulation synthesis). In FM Synthesis there is an oscillator that produces the sound signal, the <strong>Carrier</strong>, and an oscillator that modulates the carrier's wave frequency, the <strong>Modulator</strong>.

<img width="300px" src="./public/assets/fm.png">

<p><strong>Sine wave:</strong></p>
<img width="300px" src="./public/assets//waveforms/sine_wave.png">

<p><strong>Frequency-modulated sine wave:</strong></p>
<img width="300px" src="./public/assets//waveforms/sine_wave_mod.png">

<br><br>

<img width="50px" src="./screenshots/elements/synth.png">
In JSS-01, the <strong>Synth Section</strong> of the synthesizer includes the controls for the <strong>Carrier</strong> oscillator.

<br><br>

<img width="100px" src="./screenshots/elements/modulation.png">
The <strong>Modulation Section</strong> includes the controls for the <strong>Modulator</strong> oscillator.

<br><br>

<img width="58px" src="./screenshots/elements/effects.png">
The <strong>Effects Section</strong> includes the controls of the various effects that can be applied to the sound. The effects are connected in series with the following connectivity: Oscillator 2 --> Oscillator 1 --> Auto Filter --> Phaser --> Crusher --> Chebyshev --> Feedback Delay --> Ping Pong Delay --> Reverb --> Chorus --> Tremolo --> Vibrato --> Distortion --> Frequency Shifter --> Output.

#### How to Play

You can play the JSS-01 by using the on-screen keyboard, your computer keyboard, or a MIDI keyboard.

<img width="500px" src="./public/assets/keyboard/keyboard_bg_w.png">

#### Oscilloscope

The [Oscilloscope](https://en.wikipedia.org/wiki/Oscilloscope) shows the waveform of the sound. You can click on it to pause or you can right-click on it to save the current waveform as an image.

#### Modulation Index

<img width="80px" src="./screenshots/elements/mod_index.png">
The Modulation Index determines the amount of the modulation that will be applied to the Carrier. If you set the <strong>Modulation Index</strong> to 0 (can be found at the top line of the Synth section) you get the unmodulated output of the carrier oscillator.

#### MIDI

In order to use your [MIDI keyboard](https://en.wikipedia.org/wiki/MIDI_keyboard) with the JSS-01 you should connect it and turn it on, prior to accessing the web app. If you connect it while already on the JSS-01, you should refresh the page. The available MIDI devices will appear inside the MIDI display, located on top left. You can click on the MIDI device you would like to use and wait for the confirmation message. You are now ready to use your MIDI keyboard. Enjoy playing!

#### Sequencer

<img width="900px" src="./screenshots/elements/sequencer.png">
Right above the on-screen keyboard you can find the Sequencer. There is an included sequence preinstalled, so you can just press the <strong>Play button</strong> and listen to it. <br><br>

The first input from the left controls the rate of the sequence (speed). The second input is the base value of the notes of the sequence (16n stands for a 16th note value, 8n for an 8th and so on). You can input your own sequences in the main input by writing the sequence of the desired notes in this form: <strong>"C4", "F5", "D2"\*\*</strong>. You can include parts with notes of half the base value by including them in brackets, ex. <strong>"C4", ["F5", "D2"]</strong>. In these examples, if the base value is 16n (16th notes), C4 will have a value of 16th, and each of the F5 and D2 will have a value of 32th (or a value of a 16th combined).

When your sequence is ready, you can add it to the sequencer using the <strong>Add button</strong>. In order to go back to the default sequence, you can simply write <strong>default</strong> in the sequence input field and then click the Add button. There are two extra sequences included, the bassline of "I Feel Love" by Donna Summer (produced by Giorgio Moroder), which can be accessed by writing <strong>i feel love</strong> in the sequence input field and then pressing the add button, and the riff of "Funky Town", that can be accessed by writing <strong>funky town</strong>.

You can find some additional information about the JSS-01 and sound synthesis in this short [presentation](./public/assets/presentation/JSS%20Presentation.pdf).

## System Requirements

The _JSS-01 | JavaScript Software Synthesizer_ is quite heavy, especially on the CPU. We would suggest these indicative minimum requirements:

| Component  | Minimum                                 | Recommended                            |
| ---------- | --------------------------------------- | -------------------------------------- |
| CPU        | Quad-core x86-64 (2014 or later)        | Quad-core x86-64 (2018 or later) or M1 |
| RAM        | 8GB                                     | 16GB                                   |
| Display    | 1600x900                                | 1920X1080                              |
| Sound Card | On-board, dedicated, or audio interface | Audio interface                        |
| Speakers   | Monitor speakers or mixing headphones   | Monitor speakers or mixing headphones  |

#### Browser Requirements

| Browser                                                                                                                       |
| ----------------------------------------------------------------------------------------------------------------------------- |
| Chrome or other Chromium-based browsers                                                                                       |
| Safari                                                                                                                        |

CHANGELOG.md
# 3.0.0

# 2.0.3

### Changes

- Updated instructions for building and running locally

# 2.0.2

### Changes

- Migration to Vercel

# 2.0.1

### Changes

- Fixed readme images path

# 2.0.0

### Changes

- Vite migration
- Refactoring
- Code cleanup

# 1.2.0

### Changes

- Fixed computer keyboard playability (language-independent)
- Rewritten keyMapper and midiToNoteString in a concise and optimal way
- Added checks for TypeScript
- Optimized the size of all PNG and SVG images
- Code cleanup
- Updated all dependencies to their latest versions
- Fixed typos in all readme files
- Minor documentation updates
- Minor updates and corrections

# 1.0.6

### Changes

- Updated Tone.js to latest version
- Updated all dependencies to their latest versions
- Using yarn.lock in instead of package-lock
- Updated logo files
- Changed TypeScript target language to ES2021
- Readme update and restructuring
- Minor documentation updates
- Added Security Policy

# 1.0.1

### Changes

- Switched to GNU Affero General Public License v3.0

# 1.0.0

### Changes

- Migration to TypeScript

# 0.7.0

### Changes

- Dark Mode (inverted colors)

# 0.6.2

### Changes

- Updated all components to the new license
- Code cleanup

# 0.6.1

### Changes

- Converted a significant part of the codebase to TypeScript
- License changed to GNU Affero General Public License v3.0
- Minor UI fixes

# 0.5.4

### Changes

- MIDI input device selectable

### Known Issues

- When using the on-screen and/or the computer keyboard, if playing multiple notes, the synthesizer stops playing all notes if at least one note is released
- Partials sections don't work
- High CPU load and performance issues
- Opening the Dev Tools in Chrome results to severe performance deterioration: the app becomes unresponsive and crushes
- Severe performance issues on Firefox

# 0.5.3

### Changes

- Quick Start menu
- Removal of p5.js animated background for better performance
- Styling
- Quick Start documentation

### Known Issues

- When using the on-screen and/or the computer keyboard, if playing multiple notes, the synthesizer stops playing all notes if at least one note is released
- Partials sections don't work
- High CPU load and performance issues
- Opening the Dev Tools in Chrome results to severe performance deterioration: the app becomes unresponsive and crushes
- Severe performance issues on Firefox

# 0.5

### Changes

- Effects
  - Auto filter
  - Chorus
  - Phaser
  - Feedback delay
  - Ping pong delay
  - Tremolo
  - Vibrato
  - Reverb
  - Distortion
  - Bit crusher
  - Frequency shifter
  - Chebyshev waveshaping
- Responsive effects section
- Synthesizer engine refactoring
- Synthesizer effects connectivity
- Splash screen
- Background animation
- Dark mode / help mode buttons removal
- 1 extra sequence example (I Feel Love)
- Styling
- Code cleanup
- All elements as JavaScript modules

### Known Issues

- Only works on Chromium and WebKit based browsers (Chrome, Edge, Brave, Safari, Opera, GNOME Web etc.)
- When using the on-screen and/or the computer keyboard, if playing multiple notes, the synthesizer stops playing all notes if at least one note is released
- Partials sections don't work
- High CPU load

# 0.2.5

### Changes

- Sequencer
  - Note sequence input
  - Tempo
  - Note value
  - Set sequence button
  - Play button
  - Stop button
  - 2 sequence examples (default and Funky Town)
- Styling
- Code cleanup

### Known Issues

- Only works on Chromium and WebKit based browsers (Chrome, Edge, Brave, Safari, Opera, GNOME Web etc.)
- When using the on-screen and/or the computer keyboard, if playing multiple notes, the synthesizer stops playing all notes if at least one note is released
- Partials sections don't work

# 0.2.4

### Changes

- Computer keyboard playability
  - 16 notes playable (one octave + 4 notes) using keys (A-L and W-P)
  - Octave up/down (using Z,X keys)
- Amplitude envelope section
  - Attack, decay, sustain, release bars
  - Attack, decay and release curves selection
- Oscillator section
  - Oscillator type
  - Partial count
  - Partials
- Modulation Section
  - Modulation type
  - Modulation partials count
  - Modulation partials
  - Modulation envelope
    - Modulation attack, decay, sustain, release bars
    - Modulation attack, decay and release curves selection
- Complete HTML refactoring and reorganization
- Version indicator in footer
- SCSS reorganization
- All section as modules
- JavaScript code cleanup and reorganization
- Added citation file
- Readme file reorganization and expansion

### Known Issues

- Only works on Chromium and WebKit based browsers (Chrome, Edge, Brave, Safari, Opera, GNOME Web etc.)
- When using the on-screen and/or the computer keyboard, if playing multiple notes, the synthesizer stops playing all notes if at least one note is released
- Partials sections don't work

# 0.0.1

The first version!

### Features

- Polyphonic FM synthesizer engine implementing Tone.js PolySynth and FMSynth
- Oscilloscope
- Spectrogram
- Gain meter
- MIDI keyboard connectivity (limited to 2nd device in the list)
- MIDI note display

scripts/main.js
/*
 *  JSS-01 |JavaScript Software Synthesizer
 *  Copyright (c) 2022 Michael Kolesidis
 *  MIT License
 *
 */

// ---------------------------------------------------------------------
// Import Elements
// ---------------------------------------------------------------------
import Header from "./elements/header.js";
import Footer from "./elements/footer.js";
import Displays from "./elements/panels/displays.js";
import SynthSectionMain from "./elements/panels/synthSectionMain.js";

import SynthSectionAmplitudeEnvelope from "./elements/panels/synthSectionAmplitudeEnvelope.js";

import SynthSectionOscillator from "./elements/panels/synthSectionOscillator.js";

import ModulationSectionMain from "./elements/panels/modulationSectionMain.js";

import ModulationSectionModulationEnvelope from "./elements/panels/modulationSectionModulationEnvelope.js";

// ---------------------------------------------------------------------
// Import Functions
// ---------------------------------------------------------------------
import consoleIntro from "./functions/consoleIntro.js";
import midiToNoteString from "./functions/midiToNoteString.js";
import keyMapper from "./functions/keyMapper.js";

// ---------------------------------------------------------------------
// Colors
// ---------------------------------------------------------------------
const BLACK = "rgb(51, 51, 51)";
const GRAY_DARK = "rgb(180, 180, 180)";
const GRAY = "rgb(240,240,243)";
const BLUE = "rgb(1, 0, 76)";
const CYAN = "rgb(35, 178, 254)";
const GREEN = "rgb(3, 214, 146)";
const YELLOW = "rgb(254, 188, 44)";

Nexus.colors.fill = GRAY; // For all NexusUI components

// ---------------------------------------------------------------------
// Dark Mode
// ---------------------------------------------------------------------
let darkMode = false;

// function toggleDark() {
//   var element = document.body;
//   element.classList.toggle("dark-mode");
//   header.style.color = "white";
//   footer.style.color = "white";
//   midiDisplay.style.color = "white";
//   midiDisplay.style.background = "rgb(100,100,100";
//   Nexus.colors.fill = "rgb(100,100,100";
//   darkMode = true;
//   header.innerHTML = Header(darkMode);
//   let toggleDarkButton = document.createElement("button"); // recreates button in dark header
//   toggleDarkButton.innerHTML = "Dark Mode"; // recreates button in dark header
//   document.querySelector("#header").appendChild(toggleDarkButton); // recreates button in dark header
// }

// let toggleDarkButton = document.createElement("button");
// toggleDarkButton.innerHTML = "Dark Mode";
// document.querySelector("#header").appendChild(toggleDarkButton);
// // toggleDark()

// toggleDarkButton.addEventListener("click", toggleDark());

// ---------------------------------------------------------------------
// Welcome Message in Console
// ---------------------------------------------------------------------
consoleIntro();

// ---------------------------------------------------------------------
// Header
// ---------------------------------------------------------------------
const header = document.getElementById("header");
header.innerHTML = Header();

// ---------------------------------------------------------------------
// Footer
// ---------------------------------------------------------------------
const footer = document.getElementById("footer");
footer.innerHTML = Footer();

// ---------------------------------------------------------------------
// Panel Sections
// ---------------------------------------------------------------------
// Displays
const displays = document.getElementById("displays");
displays.innerHTML = Displays();

// Synth Section: Main
const synthSectionMain = document.getElementById("synth-section-main");
synthSectionMain.innerHTML = SynthSectionMain();

// Synth Section: Amplitude Envelope
const amplitudeEnvelopeSection = document.getElementById("adsr-envelope")
amplitudeEnvelopeSection.innerHTML = SynthSectionAmplitudeEnvelope();

// Synth Section: Oscillator

// Modulation Section: Main

// Modulation Section: Modulation Envelope

// Effects

// ---------------------------------------------------------------------
// MIDI Display
// ---------------------------------------------------------------------
const midiDisplay = document.getElementById("midi-display");

// ---------------------------------------------------------------------
// Oscilloscope
// ---------------------------------------------------------------------
let oscilloscope = new Nexus.Oscilloscope("#oscilloscope", {
  size: [300, 150],
});
oscilloscope.connect(Tone.getDestination());
oscilloscope.colorize("accent", BLUE);
if (darkMode) {
  oscilloscope.colorize("accent", GRAY);
}

// ---------------------------------------------------------------------
// Spectrogram
// ---------------------------------------------------------------------
let spectrogram = new Nexus.Spectrogram("#spectrogram", {
  size: [300, 150],
});
spectrogram.connect(Tone.getDestination());
spectrogram.colorize("accent", BLUE);
if (darkMode) {
  spectrogram.colorize("accent", GRAY);
}

// ---------------------------------------------------------------------
// Meter
// ---------------------------------------------------------------------
let meter = new Nexus.Meter("#meter", {
  size: [45, 150],
});
meter.connect(Tone.getDestination());
meter.colorize("accent", BLUE);
if (darkMode) {
  meter.colorize("accent", GRAY);
}

// ---------------------------------------------------------------------
// Keyboard
// ---------------------------------------------------------------------
// Nexus.colors.accent = GRAY; // dark mode
// Nexus.colors.dark = GRAY_DARK; // darl mode
// Nexus.colors.light = BLACK; // dark mode

let keyboard = new Nexus.Piano("#keyboard", {
  size: [1200, 100],
  mode: "button", // 'button', 'toggle', or 'impulse'
  lowNote: 21,
  highNote: 108,
});
keyboard.colorize("accent", GRAY_DARK); // light mode

// Keyboard resizing testing
// let keyboardResizeButton = document.createElement("button");
// keyboardResizeButton.innerHTML = "Resize";
// document.querySelector("#keyboard").appendChild(keyboardResizeButton);
// keyboardResizeButton.addEventListener("click", function() {
//   keyboard.resize(800, 50);
// });

// Makes keyboard playble both with right and left click - prevents right click context menu
// let keyboardPlaceholder = document.getElementById("keyboard");
// keyboardPlaceholder.addEventListener(
//   "contextmenu",
//   function (event) {
//     event.preventDefault();
//   },
//   false
// );

// ---------------------------------------------------------------------
// Effects
// ---------------------------------------------------------------------
// Note: Unimplemented effects are commented out

// AutoFilter .connect(autoFilter)
const autoFilter = new Tone.AutoFilter("4n").toDestination().start();
autoFilter.depth.value = 1; // range:0-1
autoFilter.frequency.value = 10; // range:0-1000 or 2000
autoFilter.octaves = 2.6; // range: -10-10

// AutoPanner

// AutoWah

// BitCrusher .connect(crusher)
const crusher = new Tone.BitCrusher(4).toDestination(); // range:1-16, step:1
crusher.bits.value;

// Chebyshev .connect(cheby)
const cheby = new Tone.Chebyshev(50).toDestination(); // range:1-100

// Chorus .connect(chorus)
const chorus = new Tone.Chorus(4, 2.5, 0.5).toDestination().start(); // frequency delayTime depth
chorus.frequency.value = 4; // range: 0-50
chorus.delayTime = 2.5; // range:0-200
chorus.depth = 0.5; // range: 0-1

// Distortion .connect(dist)
const dist = new Tone.Distortion(0.9).toDestination(); // range:0-1
dist.distortion = 0.9;

// FeedbackDelay .connect(feedbackDelay)
const feedbackDelay = new Tone.FeedbackDelay("8n", 0.5).toDestination();
feedbackDelay.delayTime.value = 0.25; // range:0-1
feedbackDelay.feedback.value = 0.5; // range:0-1

// Freeverb .connect(freeverb) {X}
// const freeverb = new Tone.Freeverb().toDestination();
// freeverb.dampening = 1000;

// FrequencyShifter .connect(shift)
const shift = new Tone.FrequencyShifter(42).toDestination(); // The incoming signal is shifted by this frequency value
shift.frequency.value = -600; // range:-600-600

// JCReverb .chain(delay, reverb) {X}
// const reverb = new Tone.JCReverb(0.4).toDestination();
// const delay = new Tone.FeedbackDelay(0.5);

// MidSideEffect
// https://tonejs.github.io/docs/14.7.77/MidSideEffect

// Phaser .connect(phaser)
const phaser = new Tone.Phaser({
  frequency: 15, // The speed of the phasing
  octaves: 5, // The octaves of the effect
  baseFrequency: 1000, // The base frequency of the filters
}).toDestination();

phaser.frequency.value; // range:0-70 (choice)
phaser.octaves = 5; // range:0-20 (choice)
phaser.baseFrequency = 1000; // range:0-1000 (choice)

// PingPongDelay .connect(PingPong)
const pingPong = new Tone.PingPongDelay("4n", 0.2).toDestination();
pingPong.delayTime.value = 2; // range:0-2 (choice)
pingPong.feedback.value = 0.2; // range:0-1

// PitchShift
// https://tonejs.github.io/docs/14.7.77/PitchShift

// Reverb .connect(reverb)
const reverb = new Tone.Reverb(1).toDestination(); // seconds - Check implementation
// https://tonejs.github.io/docs/14.7.77/Reverb - you have to wait until
reverb.decay = 1; // range:0-10 (choice)

// StereoWidener
// https://tonejs.github.io/docs/14.7.77/StereoWidener

// Tremolo .connect(tremolo)
const tremolo = new Tone.Tremolo(9, 0.75).toDestination().start(); // frequency (rate), depth
tremolo.frequency.value = 9; // range:0-100 (choice)
tremolo.depth.value = 0.75; // range:0-1

// Vibrato .connect(vibrato)
const vibrato = new Tone.Vibrato(9, 0.9).toDestination(); // frequency, depth
vibrato.frequency.value = 9; // range:0-900 (choice)
vibrato.depth.value; // range:0-1

// .connect(autoFilter).connect(crusher).connect(cheby).connect(chorus).connect(dist).connect(feedbackDelay).connect(shift).connect(phaser).connect(PingPong).connect(reverb).connect(tremolo).connect(vibrato)
// .toDestination()

// vibrato.wet.value = 1

// ---------------------------------------------------------------------
// Synthesizer
// ---------------------------------------------------------------------
let synth = new Tone.PolySynth(Tone.FMSynth).toDestination();

synth.set({
  maxPolyphony: 128,
});

const oscillator_types = ["sine", "square", "sawtooth", "triangle", "pulse"];

// ---------------------------------------------------------------------
// Volume
// ---------------------------------------------------------------------
// Dial
let volumeControl = new Nexus.Dial("#volume", {
  size: [75, 75],
  interaction: "vertical", // "radial", "vertical", or "horizontal"
  mode: "relative", // "absolute" or "relative"
  min: -40,
  max: 40,
  step: 0,
  value: -6,
});
volumeControl.colorize("accent", CYAN);

volumeControl.on("change", function (v) {
  synth.set({
    volume: v,
  });
});

// Number
let volumeNum = new Nexus.Number("#volume-num");
volumeNum.link(volumeControl);
volumeNum.colorize("accent", CYAN);

// ---------------------------------------------------------------------
// Detune
// ---------------------------------------------------------------------
// In cents - 100 cents = 8hz = 1 note - if detune 100, C4 becomes C4#, if detune 200 C4 becomes D4 and so on
// detune range : -1000-1000 (choice)

// Dial
let detuneControl = new Nexus.Dial("#detune", {
  size: [75, 75],
  interaction: "vertical", // "radial", "vertical", or "horizontal"
  mode: "relative", // "absolute" or "relative"
  min: -1000,
  max: 1000,
  step: 1,
  value: 0,
});
detuneControl.colorize("accent", CYAN);

detuneControl.on("change", function (v) {
  synth.set({
    detune: v,
  });
});

// Number
let detuneNum = new Nexus.Number("#detune-num");
detuneNum.link(detuneControl);
detuneNum.colorize("accent", CYAN);

// ---------------------------------------------------------------------
// Modulation Index
// ---------------------------------------------------------------------
// The modulation index is essentially the amound of modulation occuring. It is the ratio of the frequency of the modulating signal (mf) to the amplitude of the modulating signal (ma) – as in ma/mf.
// modulationIndex range: 0-300 (choice)

let modulationIndexControl = new Nexus.Dial("#modulation-index", {
  size: [75, 75],
  interaction: "vertical", // "radial", "vertical", or "horizontal"
  mode: "relative", // "absolute" or "relative"
  min: 0,
  max: 100,
  step: 1,
  value: 10,
});
modulationIndexControl.colorize("accent", CYAN);
// modulationIndexControl.colorize("fill", GRAY);

modulationIndexControl.on("change", function (v) {
  synth.set({
    modulationIndex: v,
  });
});

// Number;
let modulationIndexNum = new Nexus.Number("#modulation-index-num");
modulationIndexNum.link(modulationIndexControl);
modulationIndexNum.colorize("accent", CYAN);
modulationIndexNum.colorize("fill", GRAY);

// ---------------------------------------------------------------------
// Harmonicity
// ---------------------------------------------------------------------
//  Harmonicity is the ratio between the two voices. A harmonicity of 1 is no change. Harmonicity = 2 means a change of an octave.
// range: 0-20 (choice)

let harmonicityControl = new Nexus.Dial("#harmonicity", {
  size: [75, 75],
  interaction: "vertical", // "radial", "vertical", or "horizontal"
  mode: "relative", // "absolute" or "relative"
  min: 0,
  max: 20,
  step: 0,
  value: 3,
});
harmonicityControl.colorize("accent", CYAN);

harmonicityControl.on("change", function (v) {
  synth.set({
    harmonicity: v,
  });
});

// Number
let harmonicityNum = new Nexus.Number("#harmonicity-num");
harmonicityNum.link(harmonicityControl);
harmonicityNum.colorize("accent", CYAN);

// ---------------------------------------------------------------------
// ADSR Envelope
// ---------------------------------------------------------------------
// https://tonejs.github.io/docs/Envelope.html

//          /\
//         /  \
//        /    \
//       /      \
//      /        \___________
//     /                     \
//    /                       \
//   /                         \
//  /                           \

// Attack
// Range: 0 to 2
// attackCurve
// defaults: 0.01 linear

// Decay
// Range: 0+ to 2
// decayCurve
// defaults: 0.01 linear

// Sustain
// Range: 0 to 1
// The percent of the maximum value that the envelope rests at untilthe release is triggered. ()
// default: 1

// Release
// Range: 0+ to  * seconds
// releaseCurve
// defaults: 0.5 linear
// synth.options.envelope.release = 0.5;

let amplitudeADSR = new Nexus.Multislider("#amplitude-adsr", {
  size: [250, 140],
  numberOfSliders: 4,
  min: 0,
  max: 1,
  step: 0,
  candycane: 3,
  values: [0.005, 0.005, 1, 0.1],
  smoothing: 0,
  mode: "bar",
});
amplitudeADSR.colorize("accent", CYAN);

amplitudeADSR.on("change", function (v) {
  synth.set({
    envelope: {
      attack: Nexus.scale(v[0], 0, 1, 0, 2),
      decay: Nexus.scale(v[1], 0, 1, 0, 2),
      sustain: v[2],
      release: Nexus.scale(v[3], 0, 1, 0, 5),
    },
  });
});

let attackReleaseOptions = [
  "linear",
  "exponential",
  "sine",
  "cosine",
  "bounce",
  "ripple",
  "step",
];
let decayOptions = ["linear", "exponential"];

// Attack Curve
let attackCurveSelector = new Nexus.Select("#attack-curve", {
  size: [100, 30],
  options: attackReleaseOptions,
});

attackCurveSelector.on("change", function (v) {
  synth.set({
    envelope: {
      attackCurve: v.value,
    },
  });
});

// Decay Curve
let decayCurveSelector = new Nexus.Select("#decay-curve", {
  size: [100, 30],
  options: decayOptions,
});

decayCurveSelector.on("change", function (v) {
  synth.set({
    envelope: {
      decayCurve: v.value,
    },
  });
});

// Release Curve
let releaseCurveSelector = new Nexus.Select("#release-curve", {
  size: [100, 30],
  options: attackReleaseOptions,
});

releaseCurveSelector.on("change", function (v) {
  synth.set({
    envelope: {
      releaseCurve: v.value,
    },
  });
});

// ---------------------------------------------------------------------
// Modulation
// ---------------------------------------------------------------------
// partialCount
// partials []
// phase
// type*
synth.options.modulation.type; // sine, square (default), sawtooth,triangle, pulse

// ---------------------------------------------------------------------
// Modulation Envelope
// ---------------------------------------------------------------------
// attack
// attackCurve
// decay
// decayCurve
// sustain
// release
// releaseCurve

// ---------------------------------------------------------------------
// Oscillator
// ---------------------------------------------------------------------
// partialCount
// partials
// phase
// type*

synth.options.oscillator.type = "sine"; // sine, square, sawtooth,triangle, pulse

// ---------------------------------------------------------------------
// Portamento
// ---------------------------------------------------------------------
// Only available for monophonic synths

// ---------------------------------------------------------------------
// Synthesizer On-Screen Keyboard Playbility Implementation
// ---------------------------------------------------------------------
let notes = []; // For polyphonic synths
keyboard.on("change", (note) => {
  if (note.state) {
    synth.triggerAttack(midiToNoteString(note.note));
    notes.push(midiToNoteString(note.note));
  } else {
    synth.triggerRelease(notes); // Polymphinic synths need a note or an array of notes
    notes = notes.filter((e) => e !== midiToNoteString(note.note));
  }
});

// ---------------------------------------------------------------------
// Computer Keyboard Playbility Implementation
// ---------------------------------------------------------------------
let base = 39; // Middle C / C4

document.addEventListener("keydown", (event) => {
  const keyIndex = keyMapper(event.key, base);
  if (
    keyIndex >= 0 &&
    keyIndex <= 87 &&
    !keyboard.keys[keyIndex]._state.state
  ) {
    keyboard.toggleIndex(keyIndex, true);
  }
});

document.addEventListener("keyup", (event) => {
  const keyIndex = keyMapper(event.key, base);
  if (keyIndex >= 0 && keyIndex <= 87 && keyboard.keys[keyIndex]._state.state) {
    keyboard.toggleIndex(keyIndex, false);
  }
});

document.addEventListener("keydown", octaveSwitch);

function octaveSwitch(e) {
  if (e.code == "KeyZ") {
    base -= 12; // One octave down
  } else if (e.code == "KeyX") {
    base += 12; // One octave up
  }
}

// ---------------------------------------------------------------------
// MIDI Implementation
// ---------------------------------------------------------------------

// Enable WebMidi.js and trigger the onEnabled() function when ready.
WebMidi.enable()
  .then(onEnabled)
  .catch((err) => console.log(err));

function onEnabled() {
  if (WebMidi.inputs.length < 1) {
    midiDisplay.innerHTML += "No device detected.";
  } else {
    WebMidi.inputs.forEach((device, index) => {
      midiDisplay.innerHTML += `${index}: ${device.name} <br>`;
    });
  }

  const mySynth = WebMidi.inputs[1];
  // It uses input 1 by default - make it selectable by user
  // In Linux input 0 is occupied bt Midi Through Port-0
  // const mySynth = WebMidi.getInputByName("TYPE NAME HERE!")

  mySynth.channels[1].addListener("noteon", (e) => {
    synth.triggerAttack(midiToNoteString(e.data[1]));
    // notes.push(midiToNoteString(e.data[1]));
    midiDisplay.innerHTML = `<p style="font-size: 0.9rem; font-weight: 400;">MIDI note played: ${
      e.data[1]
    }<br>
    Note name: ${midiToNoteString(e.data[1])}</p>`;
  });

  mySynth.channels[1].addListener("noteoff", (e) => {
    synth.triggerRelease(midiToNoteString(e.data[1]));
  });
}

// ---------------------------------------------------------------------
// Show/Hide Section Toggle
// ---------------------------------------------------------------------
let synthSectionTitle = document.getElementById("synth-title");
let synthSectionContent = document.getElementById("synth-section-content");

let amplitudeEnvelopeTitle = document.getElementById("adsr-envelope-title");
let amplitudeEnvelope = document.getElementById("adsr-envelope");

function showHide(title, section, display) {
  // (title, section)
  title.addEventListener("click", function () {
    if (section.style.display === display) {
      section.style.display = "none";
    } else {
      section.style.display = display;
    }
  });
}

showHide(synthSectionTitle, synthSectionContent, "block");
showHide(amplitudeEnvelopeTitle, amplitudeEnvelope, "flex");
